client-runtime/src/main/java/com/microsoft/rest/v2/http/HttpPipeline.java

```diff
@@ -0,0 +1,168 @@
+     */
+    private final List<RequestPolicy.Factory> requestPolicyFactories;
+
+    /**
+     * The HttpClient that will be used to send requests unless the sendRequestAsync() method is
+     * called with a different HttpClient.
+     */
+    private HttpClientRequestPolicyAdapter httpClientRequestPolicyAdapter;
+
+    /**
+     * Create a new HttpPipeline with the provided RequestPolicy factories.
+     * @param requestPolicyFactories The RequestPolicy factories to apply to HTTP requests and
+     *                               responses that pass through this HttpPipeline.
+     * @param httpClient The HttpClient that this HttpPipeline will use to send HTTP requests.
+     */
+    private HttpPipeline(List<RequestPolicy.Factory> requestPolicyFactories, HttpClient httpClient) {
+        this.requestPolicyFactories = requestPolicyFactories;
+        this.httpClientRequestPolicyAdapter = new HttpClientRequestPolicyAdapter(httpClient);
+    }
+
+    /**
+     * Send the provided HTTP request using this HttpPipeline's HttpClient after it has passed through
+     * each of the RequestPolicies that have been configured on this HttpPipeline.
+     * @param httpRequest The HttpRequest to send.
+     * @return The HttpResponse that was received.
+     */
+    public Single<HttpResponse> sendRequestAsync(HttpRequest httpRequest) {
+        RequestPolicy requestPolicy = httpClientRequestPolicyAdapter;
+        for (final RequestPolicy.Factory requestPolicyFactory : requestPolicyFactories) {
+            requestPolicy = requestPolicyFactory.create(requestPolicy);
+        }
+        return requestPolicy.sendAsync(httpRequest);
+    }
+
+    /**
+     * Build a new HttpPipeline that will use the provided HttpClient and RequestPolicy factories.
+     * @param httpClient The HttpClient to use.
+     * @return The built HttpPipeline.
+     */
+    public static HttpPipeline build(HttpClient httpClient) {
+        return build((RequestPolicy.Factory[]) null, httpClient);
+    }
+
+    /**
+     * Build a new HttpPipeline that will use the provided HttpClient and RequestPolicy factories.
+     * @param requestPolicyFactory The RequestPolicy factory to use.
+     * @param httpClient The HttpClient to use.
+     * @return The built HttpPipeline.
+     */
+    public static HttpPipeline build(RequestPolicy.Factory requestPolicyFactory, HttpClient httpClient) {
+        return build(new RequestPolicy.Factory[] {requestPolicyFactory}, httpClient);
+    }
+
+    /**
+     * Build a new HttpPipeline that will use the provided HttpClient and RequestPolicy factories.
+     * @param requestPolicyFactory1 The first RequestPolicy factory to use.
+     * @param requestPolicyFactory2 The second RequestPolicy factory to use.
+     * @param httpClient The HttpClient to use.
+     * @return The built HttpPipeline.
+     */
+    public static HttpPipeline build(RequestPolicy.Factory requestPolicyFactory1, RequestPolicy.Factory requestPolicyFactory2, HttpClient httpClient) {
```

Comment:
You can use `build(HttpClient httpClient, RequestPolicy.Factory... factories)`. It accepts an array too.

Reply:
I changed the order of the parameters for that function. I felt that it better reflected the flow that HttpRequests would take as they go through the HttpPipeline. They start with the RequestPolicies and then end up at the HttpClient. That's the reason I added those overloads that take a fixed number of RequestPolicy.Factory objects. It's an attempt to try to make it easier to use and to visualize the pipeline.